{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Deduper part 1: planning\n",
    "\n",
    "### Purpose\n",
    "\n",
    "Filter putative PCR duplicates in a SAM file, resulting in a new file with only one read per unique library fragment. We determine a fragment's identity from: chromosome + UMI + position + strand.\n",
    "\n",
    "### Main challenges\n",
    "\n",
    " * Soft-clipping obscures \"true\" start position\n",
    "    - TACTIC: use CIGAR string to correct for soft-clipping\n",
    "    \n",
    "    \n",
    " * Start position is the left-most aligned nt, regardless of strand\n",
    "    - TACTIC: use bitwise flag to determine strand for soft-clip correction\n",
    "    \n",
    "    \n",
    " * samtools sort doesn't take soft-clipping into account, duplicates may not be adjacent in input file, file too big to store all entries in RAM\n",
    "    - TACTIC: use a dict of UMIs that each points to a set of hashes to store essential information about prior PCR duplicates\n",
    "    \n",
    "    \n",
    " * PCR dupe \"hash\" set may be too big to keep all in memory\n",
    "    - TACTIC: reset/clear \"hash\" sets whenever we get to a new chromosome/group/whatever\n",
    "    \n",
    "### Stretch goals (in order of priority)\n",
    "\n",
    " * Handle paired-end reads\n",
    "    - TACTIC: process read 1 and read 2 as before, only call PCR dupe if _both_ are dupes of the same prior molecule. This basically just means incorporating both read 1 and read 2 data into the hash.\n",
    "    \n",
    "    \n",
    " * Keep highest-quality read from each set up dupes\n",
    "    - TACTIC: Instead of a hash _set_, use a hash _dict_ with key = hash, val = (quality, read) pair. This is kind of lame, since we would then have to iterate over our sets to write reads at the end.\n",
    "    \n",
    "    \n",
    " * UMI error correction\n",
    "    - If UMI has _n_ or fewer switches from nearest good UMI, correct to good UMI.\n",
    "\n",
    "### Algorithm overview\n",
    "The main function will iterate over the lines in the SAM file. It will write all header lines to the output file. For each read, it will parse out the pertinent parts and determine strand (or unmapped). This will be used in conjunction with either the beginning or the end of the CIGAR string to \"correct\" the start position. The strand and corrected start position will be \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" keeping track of necessary imports \"\"\"\n",
    "from subprocess import call\n",
    "\n",
    "def getArgs():\n",
    "    \"\"\" use argparse to return arguments from invocation: \n",
    "        -f, --file: (str) required arg, absolute file path \n",
    "        -p, --paired: (action='store_true') optional arg, designates file is paired end (not single-end)\n",
    "        -u, --umi: (str) optional arg, designates file containing the list of UMIs (unset if randomers instead of UMIs)\n",
    "        -h, --help: optional arg, prints a USEFUL help message (see argparse docs)\n",
    "        \"\"\"\n",
    "    pass\n",
    "    \n",
    "def sortSAM(unsorted_file_name):\n",
    "    \"\"\" takes a SAM entry, calls samtools w/ params, returns the name of a sorted SAM file \"\"\"\n",
    "    \"\"\" TESTING: use unit test SAM file \"\"\"\n",
    "    # use 'call' to run samsort on unsorted_file_name w/ appropriate params\n",
    "    #return(sorted_file_name)\n",
    "    pass\n",
    "\n",
    "def buildDict(umi_list):\n",
    "    \"\"\" take a list of UMIs, return a dict of (UMI: empty set) pairs \"\"\"\n",
    "    \"\"\" TESTING: with UMI file \"\"\"\n",
    "    #return(umi_dict)\n",
    "    pass\n",
    "\n",
    "def getStrand(flag):\n",
    "    \"\"\" take a SAM bitwise flag, and return int for validity/strand (0 = not mapped, 1 or 2 = strand) \"\"\"\n",
    "    \"\"\" TESTING: 165 -> 0, 20 -> 0, 769 -> 1, 784 -> 2\"\"\"\n",
    "    if 4 & flag > 0: # checks for 'unmapped'\n",
    "        return(0)\n",
    "    if 16 & flag == 0: # checks for 'reverse strand'\n",
    "        return(1)\n",
    "    return(2)\n",
    "\n",
    "def unSoftClip(start_pos, cigar, strand):\n",
    "    \"\"\" use strand and CIGAR strand to correct for soft clip and return \"true\" start position \"\"\"\n",
    "    \"\"\" TESTING: strand1+start100+10M -> 100, strand1+start100+10S90M -> 90, strand1+start100+90M10S -> 100\"\"\"\n",
    "    \"\"\"          strand2+start100+100M -> 200, strand2+start100+10S90M -> 191, strand1+start100+90M10S -> 200\"\"\"\n",
    "    # subtranct non-maps at beginning of CIGAR string from start pos\n",
    "    # if strand == 2: # correct for left-based strand 2 start posititon\n",
    "        # add sum of CIGAR string to start pos\n",
    "    #return(corr_pos)\n",
    "    pass\n",
    "\n",
    "def makeHash(strand,start):\n",
    "    \"\"\" return unique float hash from strand, start position \"\"\"\n",
    "    #return(start+strand/10)\n",
    "    pass\n",
    "\n",
    "def parseRead(raw_read):\n",
    "    \"\"\" parses , returns chromosome, strand, corrected start pos, UMI \"\"\"\n",
    "    \"\"\" TESTING: use reads from unit test input file \"\"\"\n",
    "    global sorted_file # why keep passing this around?\n",
    "    # read line, split string into list\n",
    "    # call getStrand on bitwise flag, get strand\n",
    "    # call unSoftClip on start pos, get corrected start pos\n",
    "    #return(chrom,strand,start_pos,umi)\n",
    "    pass\n",
    "    \n",
    "def main():\n",
    "    \"\"\" DO IT ALL \"\"\"\n",
    "    # call getArgs to get invocation info\n",
    "    # call sortSAM to sort SAM file\n",
    "    # read UMI file, get list of UMIs\n",
    "    # call buildDict to build dict of (UMI : hash_set) pairs\n",
    "    # open sorted read file\n",
    "    # create + open results file\n",
    "    # set curr_chr to 'START'\n",
    "    # iterate over lines in file\n",
    "        # if header, write to file\n",
    "        # else parseRead\n",
    "            # if not mapped, next line\n",
    "            # else\n",
    "                # if new chromosome, record new one and blank hash sets\n",
    "                # if in UMI set, next line\n",
    "                # else add to UMI set, write to file\n",
    "    # close files\n",
    "    pass\n",
    "                \n",
    "if __name__ == \"__main__\": \n",
    "    main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Testing\n",
    "\n",
    "Unit test files in repo, individual function test files in function docstrings."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
